src/ripple/ledger/impl/ApplyView.cpp

```diff
@@ -0,0 +1,294 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/ledger/ApplyView.h>
+#include <ripple/basics/contract.h>
+#include <ripple/protocol/Protocol.h>
+#include <cassert>
+
+namespace ripple {
+
+std::pair<bool, std::uint64_t>
+ApplyView::dirInsert (
+    Keylet const& directory,
+    uint256 const& key,
+    bool strictOrder,
+    std::function<void (std::shared_ptr<SLE> const&)> describe)
+{
+    assert(keylet::page(directory, 0).key == directory.key);
+
+    auto root = peek(directory);
+
+    if (! root)
+    {
+        // No root, make it.
+        root = std::make_shared<SLE>(directory);
+        root->setFieldH256 (sfRootIndex, directory.key);
+        describe (root);
+
+        STVector256 v;
+        v.push_back (key);
+        root->setFieldV256 (sfIndexes, v);
+
+        insert (root);
+        return { true, 0 };
+    }
+
+    std::uint64_t page = root->getFieldU64(sfIndexPrevious);
+
+    auto node = root;
+
+    if (page)
+    {
+        node = peek (keylet::page(directory, page));
+        if (!node)
+            LogicError ("Directory chain: root back-pointer broken.");
+    }
+
+    auto indexes = node->getFieldV256(sfIndexes);
+
+    // If there's space, we use it:
+    if (indexes.size () < dirNodeMaxEntries)
+    {
+        if (!strictOrder)
+        {
+            // We can't use std::lower_bound here because
+            // existing pages may not be fully sorted.
```

Comment:
"Not fully sorted" implies there is _some_ order to to page, but we can't make any guarantees about the order at all. The pages are just "not sorted".